src/Arcus.Templates.AzureFunctions.EventHubs/SensorReadingFunction.cs

```diff
@@ -20,60 +23,99 @@ namespace Arcus.Templates.AzureFunctions.EventHubs
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="messageRouter"/> is <c>null</c>.</exception>
+        public SensorReadingFunction(IAzureEventHubsMessageRouter messageRouter)
+        {
+            Guard.NotNull(messageRouter, nameof(messageRouter), "Requires a message router instance to route incoming Azure EventHubs events through the sensor-reading processing");
+            _messageRouter = messageRouter;
+        }
+        
+        /// <summary>
+        /// Processes Azure EventHubs <paramref name="events"/>.
+        /// </summary>
+        /// <param name="events">The incoming events on the Azure EventHubs instance.</param>
+        /// <param name="propertiesArray">The array containing the set of properties for each received Azure EventHubs event.</param>
+        /// <param name="executionContext">The execution context for this Azure Functions instance.</param>
         [Function("sensor-reading")]
         public async Task Run(
-            [EventHubTrigger("sensors", Connection = "EventHubsConnectionString")] string[] messages,
+            [EventHubTrigger("sensors", Connection = "EventHubsConnectionString")] string[] events,
             Dictionary<string, JsonElement>[] propertiesArray,
-            FunctionContext context)
+            FunctionContext executionContext)
         {
-            ILogger logger = context.GetLogger<SensorReadingFunction>();
-            logger.LogInformation($"Event Hubs triggered with {messages.Length} messages");
-
-            for (var index = 0; index < messages.Length; index++)
+            ILogger logger = executionContext.GetLogger<SensorReadingFunction>();
+            logger.LogInformation("Azure EventHubs function triggered with {Length} events", events.Length);
+            
+            for (var index = 0; index < events.Length; index++)
             {
-                string message = messages[index];
+                string message = events[index];
                 Dictionary<string, JsonElement> properties = propertiesArray[index];
                 EventData data = CreateEventData(message, properties);
-
-                var messageContext = AzureEventHubsMessageContext.CreateFrom(data, "sensor-reading.servicebus.windows.net", "$Default", "sensors");
-                MessageCorrelationInfo correlationInfo = data.GetCorrelationInfo();
-
-                await _messageRouter.RouteMessageAsync(data, messageContext, correlationInfo, CancellationToken.None);
+                
+                AzureEventHubsMessageContext messageContext = data.GetMessageContext("sensor-reading.servicebus.windows.net", "sensors");
+                using (MessageCorrelationResult result = executionContext.GetCorrelationInfo(properties))
+                {
+                    await _messageRouter.RouteMessageAsync(data, messageContext, result.CorrelationInfo, CancellationToken.None);
```

Comment:
Same remark here regarding the await and WhenAll

Reply:
Added discussion: https://github.com/arcus-azure/arcus.templates/discussions/768